src/main/java/uk/gov/hmcts/reform/sendletter/tasks/SerialTaskRunner.java

```diff
@@ -0,0 +1,85 @@
+package uk.gov.hmcts.reform.sendletter.tasks;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import javax.sql.DataSource;
+
+/**
+ * Runs tasks in serial using Database locks
+ * to ensure the same task is not run concurrently.
+ */
+public final class SerialTaskRunner {
+
+    private static final Logger log = LoggerFactory.getLogger(SerialTaskRunner.class);
+    private final DataSource source;
+
+    public static SerialTaskRunner get(DataSource source) {
+        return new SerialTaskRunner(source);
+    }
+
+    private SerialTaskRunner(DataSource source) {
+        this.source = source;
+    }
+
+    /**
+     * Run the task with specified ID if no other task with matching
+     * id is already running.
+     * No error is thrown if such a task is already running, the
+     * supplied Runnable is simply not executed.
+     */
+    public void tryRun(long id, Runnable runnable) throws SQLException {
+        log.debug("Trying to lock {}", id);
+        try (Connection connection = source.getConnection()) {
+            boolean locked = false;
+            try {
+                if (tryLock(id, connection)) {
+                    log.debug("Acquired lock {}", id);
+                    locked = true;
+                    runnable.run();
+                } else {
+                    log.debug("Failed to acquired lock {}", id);
```

Comment:
It would be best if we could find this entry in production logs, e.g. in case we lock ourselves forever. It would also be good to have some sort of enum, so that the Id would have a corresponding task name.

Reply:
Definitely, would log.info show up in our logs?

Reply:
Yes, it would

Reply:
The task ids will be in an enum so we will be able to tie them back to task names.

Reply:
would be much more convenient for everyone to log task name imho.
Perhaps on a next PR, when this helper is actually used (and tasks/jobs defined)

Reply:
Yes, I agree